### PR TITLE
Add XBox Controller Demo

### LINKED ISSUE #322: Add ROS node to read XBox controller commands
### Description of the task

Add a ROS node that will allow us to receive commands from a connected XBox controller.

### Outline the task in more detail

We want to be able to receive commands from an XBox controller, so that we can later use them to control the robots or perform other special commands.

Luckily, ROS already has a package that should be able to do this, called `Joy`. http://wiki.ros.org/joy

This package should be able to be installed as a `rosdep`. Our `environment_setup/setup_software.sh` script will need to be modified to install the package. See https://github.com/UBC-Snowbots/Snowflake/blob/master/setup_scripts/install_dependencies.sh for an example.

The package also may need to be listed as a dependency in our `src/thunderbots/software/package.xml` file. Again, see https://github.com/UBC-Snowbots/Snowflake/blob/master/src/sb_gazebo/package.xml for an example of `gazebo` listed as a dependency.

According to Google, you _might_ need to install/use a different xbox driver. Hopefully this isn't the case, but if things don't seem to be working try another driver like https://andrewdai.co/xbox-controller-ros.html

Once you have the `joy` node up and running, you'll need to manually verify that we are receiving data properly. We have XBox controllers you can borrow for this. We don't need to connect the data to anything, we just want to make sure we are receiving it and that the `joy` node is publishing properly.

Finally, make a new launchfile that runs the `joy` node, in `src/thunderbots/software/launch/`

### Acceptance criteria

- [ ] Install the `joy` ROS node and add it as a dependency
- [ ] Make a new launchfile that launched the `joy` node
- [ ] Verify that we are able to successfully receive data from the XBox controller and that the `joy` node is publishing messages as expected

### PR DESCRIPTION
**Pull request by MathewMacDougall**
### Description

* Added a new node to read commands from connected XBox controllers
* Added a new node to convert the controller commands to Primitives

Robots can now be controlled with the XBox controller

### Testing Done

Manual testing with grSim

### Resolved Issues

resolves #322 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Start of document comments**: each `.cpp` and `.h` file should have a comment at the start of it. See files in the `thunderbots/software/geom` folder for examples.
- [x] **Function comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the classes defined in `thunderbots/software/geom`
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

**Comment by MathewMacDougall** [comment]
@akhilveeraghanta it seems like the Dynamic Parameters aren't being updated live in the code, despite them showing up in the `rqt_reconfigure` GUI and the `test_parameter_exists.test` test passing. Any ideas?

File: src/thunderbots/software/xbox_controller_mapping/main.cpp [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+        toggleDribble(static_cast<unsigned int>(buttons[Y])) *
+        Util::DynamicParameters::XBoxControllerDemo::dribbler_rpm.value();
+
+    // The triggers report values in the range [-1.0, 1.0]. Values are < 0 when the
+    // trigger is pressed/held down more than halfway, and > 0 otherwise
+    double robot_kick_speed_meters_per_second =
+        (axes[RIGHT_TRIGGER] < 0.0)
+            ? Util::DynamicParameters::XBoxControllerDemo::kick_speed_meters_per_second
+                  .value()
+            : 0.0;
```
Comment (garethellis0) [inline src/thunderbots/software/xbox_controller_mapping/main.cpp]: Could we change this to something like:
```
   double robot_kick_speed_meters_per_second = Util::DynamicParameters::XBoxControllerDemo::kick_speed_meters_per_second.value();
   if (axes[RIGHT_TRIGGER] < 0.0) {
        robot_kick_speed_meters_per_second = 0;
   }
```
and similarly below?

**Review by garethellis0 — CHANGES_REQUESTED** [review]
looks pretty reasonable to me, just one comment

File: src/thunderbots/software/xbox_controller_mapping/main.cpp (hunk 1 above)
Comment (MathewMacDougall) [reply src/thunderbots/software/xbox_controller_mapping/main.cpp]: fixed

Comment (garethellis0) [reply src/thunderbots/software/xbox_controller_mapping/main.cpp]: Doesn't look like this has changed?

**Comment by garethellis0** [comment]
Couple things:
- what's the status on the `TODO` items in the PR description? If they're not done, we could probably just open issues for them so this can get merged?
- looks like there are some merge conflicts

File: src/thunderbots/software/xbox_controller_mapping/main.cpp (hunk 1 above)
Comment (MathewMacDougall) [reply src/thunderbots/software/xbox_controller_mapping/main.cpp]: Sorry, pushed now

**Comment by MathewMacDougall** [comment]
Sorry I commented before I was finished. Changes are now pushed.

Both `TODOs` from the description are resolved (shooting and chipping now works, Akhil is making a PR for the DynamicParameters)

File: src/thunderbots/software/CMakeLists.txt [hunk 2]
```diff
@@ -224,6 +224,29 @@ target_link_libraries(parameters  ${catkin_LIBRARIES}
...
+        ${CMAKE_CURRENT_SOURCE_DIR}/util/timestamp.cpp
+        ${CMAKE_CURRENT_SOURCE_DIR}/util/ros_messages.cpp
+        )
+add_executable (xbox_controller_mapping
+        ${XBOX_CONTROLLER_MAPPING_SRC}
+        ${SHARED_UTIL_SRC}
+        )
+# Depend on exported targets (other packages) so that the messages in our thunderbots_msgs package are built first.
+# This way the message headers are always generated before they are used in compilation here.
+add_dependencies(parameters thunderbots_gencfg ${catkin_EXPORTED_TARGETS})
```
Comment (garethellis0) [inline src/thunderbots/software/CMakeLists.txt]: I think this is what's breaking CI:
``` suggestion
add_dependencies(xbox_controller_mapping ${catkin_EXPORTED_TARGETS})
```

**Review by garethellis0 — DISMISSED** [review]
LGTM :+1: :tada:

**Comment by garethellis0** [comment]
Looks like some unit tests failed in CI

**Review by garethellis0 — APPROVED** [review]
LGTM :+1: :tada:

**Pull request merged**